

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Assignment details &mdash; bioinformatics@nau</title>
    
    <link rel="stylesheet" href="../../static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="../../static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../static/print.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../static/jquery.js"></script>
    <script type="text/javascript" src="../../static/underscore.js"></script>
    <script type="text/javascript" src="../../static/doctools.js"></script>
    <script type="text/javascript" src="../../static/theme_extras.js"></script>
    <link rel="top" title="bioinformatics@nau" href="../../index.html" />
    <link rel="up" title="Bio/CS 299 - Bioinformatics I: Introduction to Bioinformatics (Fall 2011)" href="index.html" />
    <link rel="next" title="Application presentations" href="application_presentations.html" />
    <link rel="prev" title="Books and recommended references" href="books_and_links.html" /> 
  </head>
  <body>
      <div class="header"><h1 class="heading"><a href="../../index.html">
          <span>bioinformatics@nau</span></a></h1>
        <h2 class="heading"><span>Assignment details</span></h2>
      </div>
      <div class="topnav">
      
        <p>
        «&#160;&#160;<a href="books_and_links.html">Books and recommended references</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="application_presentations.html">Application presentations</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="assignment-details">
<span id="id1"></span><h1>Assignment details<a class="headerlink" href="#assignment-details" title="Permalink to this headline">¶</a></h1>
<div class="section" id="programming-assignment-3-assembling-an-overlap-table">
<h2>Programming assignment 3: Assembling an overlap table<a class="headerlink" href="#programming-assignment-3-assembling-an-overlap-table" title="Permalink to this headline">¶</a></h2>
<p>In this assignment you will assemble an overlap table that could be used in the de novo assembly of a genome from a set of short sequencing reads.</p>
<p>Create a new python file named <tt class="docutils literal"><span class="pre">prog3_&lt;userid&gt;.py</span></tt>, where <tt class="docutils literal"><span class="pre">userid</span></tt> is replaced with your NAU user id. You must also define <tt class="docutils literal"><span class="pre">userid=&lt;userid&gt;</span></tt> (again where &lt;userid&gt; is replaced with your NAU user id) in your script so that it can be imported into other code. Any deviations from these instructions <strong>will result in a zero on this assignment</strong>!</p>
<dl class="docutils">
<dt>In this file, define a function that takes a list of (sequence id, sequence) tuples and a substitution matrix. This function should return:</dt>
<dd><ul class="first last simple">
<li><tt class="docutils literal"><span class="pre">overlap_matrix</span></tt>: a 2D list where each internal list <tt class="docutils literal"><span class="pre">i</span></tt>  represents a vector of the number of overlapping positions between sequence <tt class="docutils literal"><span class="pre">i</span></tt> and every sequence in the input sequence collection;</li>
<li><tt class="docutils literal"><span class="pre">repeat_counts</span></tt>: a list containing the number of times each sequence is repeated in its entirety.</li>
</ul>
</dd>
</dl>
<p>The overlap score should be <tt class="docutils literal"><span class="pre">-1</span></tt> for the overlap between a sequence and itself or any other identical repeat. (For identical repeat I use the definition of python string equality, so if <tt class="docutils literal"><span class="pre">a</span> <span class="pre">==</span> <span class="pre">b</span></tt>, <tt class="docutils literal"><span class="pre">a</span></tt> is an identical repeat of <tt class="docutils literal"><span class="pre">b</span></tt> and vice versa.)</p>
<p>The function should be of the form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">compute_overlaps</span><span class="p">(</span><span class="n">seqs</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">):</span>
    <span class="c"># do a bunch of work</span>
    <span class="k">return</span> <span class="n">overlap_matrix</span><span class="p">,</span> <span class="n">repeat_counts</span>
</pre></div>
</div>
<p>You should additionally define a 2D dictionary <tt class="docutils literal"><span class="pre">assembly_substitution_matrix</span></tt> that contains scores for substitutions between the nucleotide bases. You will define the values in this martix to optimize them for this task.</p>
<p>Additionally, I want you to benchmark your script against several large collections of sequences, and plot the run time of each. You can get the run time by using the <tt class="docutils literal"><span class="pre">time</span></tt> command which is built-in to Unix/Linux and Mac OS X. If you are running Windows on your laptop you can either do this portion on the Amazon Cloud or in the Computer Science lab in the engineering building. It may be useful for you to generate a command line interface for your script to achieve this portion of the assignment. The <tt class="docutils literal"><span class="pre">MinimalFastaParser</span></tt> function in PyCogent will additionally be useful here: given an open fasta file, this returns a list (really a generator) of (sequence_id, sequence) tuples, which is the input expected by <tt class="docutils literal"><span class="pre">compute_overlaps</span></tt>.</p>
<p>Download the files here for your benchmark analysis <tt class="xref download docutils literal"><span class="pre">here</span></tt> For this component of the assignment, I want you to turn in a scatter plot where the x axis is the number of sequences in the input file and the y axis is the runtime of your script in seconds.</p>
<dl class="docutils">
<dt>Hints:</dt>
<dd><ul class="first last simple">
<li>You can optionally re-use the <tt class="docutils literal"><span class="pre">sw_align</span></tt> function generated in programming assignment 1. So everyone has access to a single implementation, you can use the version included <tt class="xref download docutils literal"><span class="pre">here</span></tt> To use this, place the file in the same directory as your <tt class="docutils literal"><span class="pre">prog3_&lt;userid&gt;.py</span></tt> and add the line <tt class="docutils literal"><span class="pre">from</span> <span class="pre">sw_align</span> <span class="pre">import</span> <span class="pre">sw_align</span></tt> to your script. If you choose not to use this function, be sure to include any necessary code in your <tt class="docutils literal"><span class="pre">prog3_&lt;userid&gt;.py</span></tt> file &#8211; no other imports (unless from the python standard library or PyCogent) will be allowed (i.e., it needs to run on my laptop).</li>
<li>Your substitution matrix should be optimized so mismatches are scored so poorly that in practice they never show up.</li>
<li>You&#8217;ll want to impose a high gap penalty so in practice gaps never show up.</li>
<li>Think about things you can do to optimize your code to reduce the run time. This may come up on the final...</li>
</ul>
</dd>
</dl>
<p><tt class="xref download docutils literal"><span class="pre">This</span> <span class="pre">script</span></tt> will allow you to test your code. This is a partial set of the tests that I will run. Passing all 4 of these tests will guarantee you at least a 40% on this homework assignment. After downloading this script to the same directory as your homework assignment you can run the tests as follows:</p>
<div class="highlight-python"><pre>python test_prog3.py prog3_jgc53.py</pre>
</div>
<p>(Where my script is named <tt class="docutils literal"><span class="pre">prog3_jgc53.py</span></tt>.)</p>
</div>
<div class="section" id="programming-assignment-2-constructing-a-phylogenetic-tree">
<h2>Programming assignment 2: Constructing a phylogenetic tree<a class="headerlink" href="#programming-assignment-2-constructing-a-phylogenetic-tree" title="Permalink to this headline">¶</a></h2>
<p>In this assignment you will make use of the PyCogent software package to automate the process of constructing a phylogenetic tree from a set of genes.</p>
<p>Create a new python file named <tt class="docutils literal"><span class="pre">prog2_&lt;userid&gt;.py</span></tt>, where <tt class="docutils literal"><span class="pre">userid</span></tt> is replaced with your NAU user id. You must also define <tt class="docutils literal"><span class="pre">userid=&lt;userid&gt;</span></tt> (again where &lt;userid&gt; is replaced with your NAU user id) in your script so that it can be imported into other code. Any deviations from these instructions <strong>will result in a zero on this assignment</strong>!</p>
<p>In this file, define a function that takes a list of queries to be run against NCBI and a list of query labels to label the results of each query in the final tree, and returns a phylogenetic tree derived from aligned representatives of those queries. This function should optionally take a parameter <tt class="docutils literal"><span class="pre">n</span></tt> which defines how many randomly chosen representative sequences should be chosen for each of the queries. The default value for <tt class="docutils literal"><span class="pre">n</span></tt> should be 5. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">obtain_sequences_and_build_tree</span><span class="p">(</span><span class="n">queries</span><span class="p">,</span><span class="n">query_labels</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">):</span>
    <span class="c"># do a bunch of work</span>
    <span class="k">return</span> <span class="n">tree</span>
</pre></div>
</div>
<p>You should filter any sequences that have one or more <tt class="docutils literal"><span class="pre">N</span></tt> characters in them, and each sequence id should begin with the query label corresponding to that sequence.</p>
<p><tt class="docutils literal"><span class="pre">tree</span></tt> should be a PyCogent <tt class="docutils literal"><span class="pre">PhyloNode</span></tt> object. The function handle must look exactly like the one above (remember - the tests are automated so deviations from that will result in a zero!).</p>
<dl class="docutils">
<dt>Items to turn in:</dt>
<dd><ul class="first last">
<li><p class="first">Your <tt class="docutils literal"><span class="pre">.py</span></tt> file, by e-mail.</p>
</li>
<li><p class="first">Apply your function as follows and e-mail me a newick formatted tree file named <tt class="docutils literal"><span class="pre">prog2_&lt;userid&gt;.tre</span></tt>, where <tt class="docutils literal"><span class="pre">userid</span></tt> is replaced with your NAU user id. Be sure your tree includes distances (i.e., branch lengths):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">obtain_sequences_and_build_tree</span><span class="p">(</span>
     <span class="p">[</span><span class="s">&#39;&quot;small subunit rRNA&quot;[ti] AND archaea[orgn]&#39;</span><span class="p">,</span>
      <span class="s">&#39;&quot;small subunit rRNA&quot;[ti] AND bacteria[orgn]&#39;</span><span class="p">,</span>
      <span class="s">&#39;&quot;small subunit rRNA&quot;[ti] AND eukarya[orgn]&#39;</span><span class="p">],</span>
     <span class="p">[</span><span class="s">&#39;A: &#39;</span><span class="p">,</span><span class="s">&#39;B: &#39;</span><span class="p">,</span><span class="s">&#39;E: &#39;</span><span class="p">],</span>
     <span class="n">n</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
</pre></div>
</div>
</li>
</ul>
</dd>
<dt>Hints:</dt>
<dd><ul class="first last simple">
<li><a class="reference external" href="http://dl.dropbox.com/u/2868868/pycogent_160dev_docs/cookbook/building_a_tree_of_life.html">This page</a> will help.</li>
<li>The Amazon QIIME EC2 instance has PyCogent, muscle, and FastTree preinstalled. Working there will save you a lot of time on software installation.</li>
<li>Remember that you can call <tt class="docutils literal"><span class="pre">dir()</span></tt> on an object to find out what methods are available to that object. One of the methods associated with your tree object will help you generate a newick formatted tree.</li>
</ul>
</dd>
</dl>
<p><tt class="xref download docutils literal"><span class="pre">This</span> <span class="pre">script</span></tt> will allow you to test your code. This is a partial set of the tests that I will run. Passing all 5 of these tests will guarantee you at least a 50% on this homework assignment. After downloading this script to the same directory as your homework assignment you can run the tests as follows:</p>
<div class="highlight-python"><pre>python test_prog2.py prog2_jgc53.py</pre>
</div>
<p>(Where my script is named <tt class="docutils literal"><span class="pre">prog2_jgc53.py</span></tt>.)</p>
</div>
<div class="section" id="programming-assignment-1-i-e-modified-exercise-41-44-extra-credit">
<h2>Programming assignment 1 (i.e., Modified exercise 41-44): Extra credit<a class="headerlink" href="#programming-assignment-1-i-e-modified-exercise-41-44-extra-credit" title="Permalink to this headline">¶</a></h2>
<p>Create a new python file named <tt class="docutils literal"><span class="pre">prog1ec_&lt;userid&gt;.py</span></tt>, where <tt class="docutils literal"><span class="pre">userid</span></tt> is replaced with your NAU user id. For example, mine is called <tt class="docutils literal"><span class="pre">prog1ec_jgc53.py</span></tt>. Name it exactly like this (including case!). You must also define <tt class="docutils literal"><span class="pre">userid=&lt;userid&gt;</span></tt> (again where &lt;userid&gt; is replaced with your NAU user id) in your script so that it can be imported into other code. Any deviations from these instructions <strong>will result in a zero on this assignment</strong>!</p>
<p>Define a function of the form:</p>
<div class="highlight-python"><pre>def transcribe(dna)</pre>
</div>
<p>which simulates transcription of a string passed as <tt class="docutils literal"><span class="pre">dna</span></tt> and returns the resulting sequence as a string.</p>
<p>Define another function of the form:</p>
<div class="highlight-python"><pre>def rc(dna)</pre>
</div>
<p>which reverse complements a string passed as <tt class="docutils literal"><span class="pre">dna</span></tt> and returns the resulting sequence as a string.</p>
<p>Your python file should also define a command line interface that allows users to pass a single sequence followed by an operation (<tt class="docutils literal"><span class="pre">t</span></tt> for transcribe or <tt class="docutils literal"><span class="pre">rc</span></tt> for reverse complement). The script should apply the appropriate operation and print the result to the terminal. If the user passes something other than <tt class="docutils literal"><span class="pre">t</span></tt> or <tt class="docutils literal"><span class="pre">rc</span></tt>, for example <tt class="docutils literal"><span class="pre">xyz</span></tt>, the script should print <tt class="docutils literal"><span class="pre">Unknown</span> <span class="pre">operation:</span> <span class="pre">xyz.</span></tt></p>
<p>You should be able to call your script as follows:</p>
<div class="highlight-python"><pre>python prog1ec_jgc53.py ACCGT t
python prog1ec_jgc53.py ACCGT rc</pre>
</div>
<p>Validate your script by downloading <tt class="xref download docutils literal"><span class="pre">this</span> <span class="pre">script</span></tt> by saving it to the same directory as your script and running the command:</p>
<div class="highlight-python"><pre>python test_prog1ec.py prog1ec_jgc53.py</pre>
</div>
<p>When your script is working correctly the output should end with:</p>
<div class="highlight-python"><pre>----------------------------------------------------------------------
Ran 5 tests in 0.117s

OK</pre>
</div>
<p>There are five tests implemented here, and each test is worth one extra credit point toward programming assignment 1. These are the same tests that I will run when you turn your assignment in.</p>
</div>
<div class="section" id="modified-exercise-41-44">
<h2>Modified exercise 41-44<a class="headerlink" href="#modified-exercise-41-44" title="Permalink to this headline">¶</a></h2>
<p>This exercise will be broken apart for the CS students and the Bio students. You should start this exercise with my Needleman-Wunsch python implementation presented here. If you feel that you should be doing the assignment assigned to the other group, e-mail me by Friday Oct 7th at the latest with an explanation of why and I will consider your request.</p>
<p>Begin with the Needleman-Wunsch implementation in the <a class="reference external" href="../lectures/lecture11.html">lecture 11 notes</a>.</p>
<div class="section" id="all-students">
<h3>All students<a class="headerlink" href="#all-students" title="Permalink to this headline">¶</a></h3>
<p>Add a new function with this <cite>exact</cite> form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">nw_align</span><span class="p">(</span><span class="n">sequence1</span><span class="p">,</span><span class="n">sequence2</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">)</span>
</pre></div>
</div>
<p>This function should return, in this order, the aligned sequence 1 as a string, the aligned sequence 2 as a string, and the score of the global alignment.</p>
<p>Next, add a command line interface so users can call this with arbitrary sequences from the command line and have the aligned sequences and the score print to the screen.</p>
<p>Save this script named as follows replacing <tt class="docutils literal"><span class="pre">&lt;userid&gt;</span></tt> with your NAU user id <tt class="docutils literal"><span class="pre">prog1_&lt;userid&gt;.py</span></tt>. For example, my file would be called <tt class="docutils literal"><span class="pre">prog1_jgc53.py</span></tt> because my NAU id is <tt class="docutils literal"><span class="pre">jgc53</span></tt>.</p>
<p>Toward the top of the file define a variable <tt class="docutils literal"><span class="pre">userid</span></tt> that is set to your NAU user id. This is <cite>very important</cite> - you will receive a zero on this assignment if you do not do this. For me, this would look like:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">userid</span> <span class="o">=</span> <span class="s">&quot;jgc53&quot;</span>
</pre></div>
</div>
<p>To confirm that your script is formatted correctly, <tt class="xref download docutils literal"><span class="pre">this</span> <span class="pre">script</span></tt>. Place this script in the same directory as your homework assignment and run the following command:</p>
<div class="highlight-python"><pre>validate_assignment_specifications.py &lt;script_name&gt;</pre>
</div>
<p>Where <tt class="docutils literal"><span class="pre">&lt;script_name&gt;</span></tt> is replaced by the name of your script. For example, my command would look like:</p>
<div class="highlight-python"><pre>validate_assignment_specifications.py prog1_jgc53.py</pre>
</div>
<p>You should get a message that three tests passed. If the message does not end with the following you have not formatted your homework correctly, and you will get a zero on the assignment:</p>
<div class="highlight-python"><pre>----------------------------------------------------------------------
Ran 3 tests in 0.007s

OK</pre>
</div>
</div>
<div class="section" id="bio-students">
<h3>Bio students<a class="headerlink" href="#bio-students" title="Permalink to this headline">¶</a></h3>
<p>In the script you created above, define a new function of the form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">generate_random_score_distribution</span><span class="p">(</span><span class="n">query_sequence</span><span class="p">,</span><span class="n">subject_sequence</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">)</span>
</pre></div>
</div>
<p>Which returns a list of <tt class="docutils literal"><span class="pre">n</span></tt> scores for aligning each of <tt class="docutils literal"><span class="pre">n</span></tt> random sequences of the same length as <tt class="docutils literal"><span class="pre">query_sequence</span></tt> against <tt class="docutils literal"><span class="pre">subject_sequence</span></tt>.</p>
<p>Next, define a function that takes a query sequence, a subject sequence, and a value <tt class="docutils literal"><span class="pre">n</span></tt> with this form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">fraction_better_or_equivalent_alignments</span><span class="p">(</span><span class="n">query_sequence</span><span class="p">,</span><span class="n">subject_sequence</span><span class="p">,</span><span class="n">n</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">)</span>
</pre></div>
</div>
<p>This function should call <tt class="docutils literal"><span class="pre">generate_random_score_distribution</span></tt> to generate a list of scores for random alignments. It should then compute the score for aligning <tt class="docutils literal"><span class="pre">query_sequence</span></tt> against <tt class="docutils literal"><span class="pre">subject_sequence</span></tt>. The return value of this function should be the number of random alignment scores that are better or equal to the actual alignment score divided by <tt class="docutils literal"><span class="pre">n</span></tt>.</p>
<p>Generate a length 100 subject sequence randomly. Next generate about 5 query sequences of varied lengths (around 3 to around 75), and log the results of comparing each query sequence to the subject sequence. Log the sequences and results in your programming notebook. Describe any general patterns that you notice. What do you think this means? Run this example on the alignment we worked through in class (query sequence: <tt class="docutils literal"><span class="pre">HEAGAWGHEE</span></tt>; subject sequence: <tt class="docutils literal"><span class="pre">PAWHEAE</span></tt>) and describe the results. How does this alignment compare to your randomly generated alignments? Justify your answer in your programming notebook.</p>
<dl class="docutils">
<dt>Hints:</dt>
<dd><ul class="first last simple">
<li>In the Lecture 8 notes there is code illustrating how to generate a random sequence of bases at a given sequence length. Here we&#8217;re working with protein sequences, but the process is the same.</li>
<li>Running your fraction_better_alignments on the query and subject sequences below will go quicker if you hook this up to the command line interface of your script. Every time the script is called, run this analysis and include the results at the end.</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="cs-students">
<h3>CS students<a class="headerlink" href="#cs-students" title="Permalink to this headline">¶</a></h3>
<p>You will define a general function that can perform global (Needleman-Wunsch; NW) or local (Smith-Waterman; SW) alignments.</p>
<p>Define a new function, <tt class="docutils literal"><span class="pre">generate_sw_and_traceback_matrices</span></tt> with the following form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">generate_sw_and_traceback_matrices</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">,</span><span class="n">gap_penalty</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">)</span>
</pre></div>
</div>
<p>The return value should be the dynamic programming matrix and the traceback matrix for a SW alignment.</p>
<p>Define a new function <tt class="docutils literal"><span class="pre">sw_traceback</span></tt> with the form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">sw_traceback</span><span class="p">(</span><span class="n">traceback_matrix</span><span class="p">,</span><span class="n">sw_matrix</span><span class="p">,</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">)</span>
</pre></div>
</div>
<p>This function should return aligned the aligned sequences in the order they were passed in and the alignment score.</p>
<p>Define a new function <tt class="docutils literal"><span class="pre">align</span></tt> with the following form:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">align</span><span class="p">(</span><span class="n">sequence1</span><span class="p">,</span><span class="n">sequence2</span><span class="p">,</span><span class="n">substitution_matrix</span><span class="p">,</span><span class="n">local</span><span class="p">)</span>
</pre></div>
</div>
<p>Where <tt class="docutils literal"><span class="pre">local</span></tt> is a boolean value. This function should return aligned_sequence1, aligned_sequence2, and the best alignment score. If <tt class="docutils literal"><span class="pre">local==False</span></tt>, an NW alignment should be performed. If <tt class="docutils literal"><span class="pre">local==True</span></tt> an SW alignment should be performed.</p>
<p>Hook this up to the command line interface so the user can optionally create a global or a local alignment.</p>
<dl class="docutils">
<dt>Hints:</dt>
<dd><ul class="first last simple">
<li>In my <a class="reference external" href="http://dl.dropbox.com/u/2868868/cs299_slides_XCFGcsdFGGad/Lecture9.pptx.pdf">slides from lecture 9</a> I provide details on the differences between SW and NW initialization, scoring, and traceback.</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="modified-exercise-38">
<h2>Modified exercise 38<a class="headerlink" href="#modified-exercise-38" title="Permalink to this headline">¶</a></h2>
<p>For LPTHW modified exercise 38 I want you to find and review bioinformatics code (rather than some random code out on the internet).</p>
<p>For folks who are already comfortable with Python, start with PyCogent, QIIME, or any other bioinformatics python code you come across. Start <a class="reference external" href="http://pycogent.svn.sourceforge.net/viewvc/pycogent/trunk/cogent/">here for PyCogent</a> or <a class="reference external" href="http://qiime.svn.sourceforge.net/viewvc/qiime/trunk/qiime/">here for QIIME</a> (after clicking on a file name, click the <tt class="docutils literal"><span class="pre">View</span></tt> link under <tt class="docutils literal"><span class="pre">Links</span> <span class="pre">to</span> <span class="pre">HEAD</span></tt>). Record in your programming notebook which file(s) you work with.</p>
<p>For students who are just learning python, review in detail the sequence evolution script that we built in Lecture 8. Work through this until you can explain what <cite>each line</cite> is doing. Ideally work with a friend and explain the lines to each other. Then make some modifications to change the behavior. You can find that <a class="reference internal" href="lectures/lecture8.html#sequence-evolution-script"><em>here</em></a>.</p>
</div>
<div class="section" id="blasting-through-the-kingdom-of-life">
<h2>BLASTing Through the Kingdom of Life<a class="headerlink" href="#blasting-through-the-kingdom-of-life" title="Permalink to this headline">¶</a></h2>
<p><strong>This assignment will be due in class on Tuesday Sept 27th. This is a graded homework assignment that must be turned in on paper (i.e., not digitally). Your answers must be typed, not hand-written, and formatted as in the example answers in</strong> <a class="reference download internal" href="../../_downloads/BTTKOL_students.pdf"><tt class="xref download docutils literal"><span class="pre">this</span> <span class="pre">document</span></tt></a> <strong>. Your name must appear at the top left of the first page.</strong> I will deduct points for deviations from these requirements.</p>
<p>In class on Tuesday Sept 20th I&#8217;ll give out assignments of which students will work with which two sequences. Everyone should work independently on this project: it&#8217;s OK to talk with other students about the project, but <strong>it is not OK look at each others results</strong>.</p>
<p>This assignment will go faster if you start by reviewing <a class="reference external" href="http://www.digitalworldbiology.com/BLAST/index.html">this tutorial</a>.</p>
<p>You can find your sequences <a class="reference external" href="http://www.digitalworldbiology.com/BLAST/62000sequences.html">here</a> and the list of assigned sequences <a class="reference download internal" href="../../_downloads/BLASTing_assignments.txt"><tt class="xref download docutils literal"><span class="pre">here</span></tt></a> .</p>
<p>You should answer the questions posted <a class="reference download internal" href="../../_downloads/BTTKOL_students.pdf"><tt class="xref download docutils literal"><span class="pre">here</span></tt></a> for each of your two sequences. Note that there are sample answers for an example sequence - you should format your answers as you see in the sample answers in this PDF (don&#8217;t forget the <cite>Information Source</cite>!).</p>
</div>
<div class="section" id="modified-exercise-36-adding-error-checking-to-our-sequence-processing-script">
<h2>Modified Exercise 36: Adding error checking to our sequence processing script<a class="headerlink" href="#modified-exercise-36-adding-error-checking-to-our-sequence-processing-script" title="Permalink to this headline">¶</a></h2>
<p>Read through Exercise 36, but don&#8217;t do the <cite>Homework</cite> section. Instead, do the following.</p>
<p>Start with the example script from lecture 5 (<a class="reference internal" href="lectures/lecture5.html#sequence-processing-script"><em>see here</em></a>: use the last one included in lecture 5 with the <tt class="docutils literal"><span class="pre">reverse_complement</span></tt> and <tt class="docutils literal"><span class="pre">transcribe</span></tt> functions.)</p>
<p>We discussed several times that the <tt class="docutils literal"><span class="pre">reverse_complement</span></tt> function as implemented is wrong. Use a <tt class="docutils literal"><span class="pre">for</span></tt> loop to improve this: this script should print an error message to the terminal and exit without writing the output file if the input sequence contains any characters other than <tt class="docutils literal"><span class="pre">A</span></tt>, <tt class="docutils literal"><span class="pre">C</span></tt>, <tt class="docutils literal"><span class="pre">G</span></tt>, or <tt class="docutils literal"><span class="pre">T</span></tt>.</p>
<p>Advanced python users: try reimplementing this function with a dictionary and with string&#8217;s <tt class="docutils literal"><span class="pre">translate</span></tt> method. Which implementation is most efficient if you try to process a lot (e.g., 100) very large strings (e.g., length 1,000,000)? Back up your answer with data (hint: check out the <tt class="docutils literal"><span class="pre">time</span></tt> command) and log your results in your programming notebook. Check out the <tt class="docutils literal"><span class="pre">choice</span></tt> method in the <tt class="docutils literal"><span class="pre">random</span></tt> module - use this with a <tt class="docutils literal"><span class="pre">for</span></tt> loop to generate your string. Can you come up with any more efficient mechanisms for reverse complementing a string (I don&#8217;t have a specific example in mind here...)? What are some reasons why you might choose one implementation over another?</p>
<p>Log your modified script in your programming notebook by pasting the code in.</p>
<p><strong>E-mail me your programming notebook before class begins on Tuesday!</strong></p>
</div>
<div class="section" id="programming-noteboook-on-going-will-be-collected-several-times-through-out-the-semester">
<h2>Programming noteboook (on-going; will be collected several times through-out the semester)<a class="headerlink" href="#programming-noteboook-on-going-will-be-collected-several-times-through-out-the-semester" title="Permalink to this headline">¶</a></h2>
<p>The programming notebook is an electronic journal to keep track of your progress with learning python. The things that I&#8217;m looking for are notes on what you had difficulties with, what you had to look up along the way, and how you solved certain problems. As you move through the exercises you&#8217;ll undoubtedly run into the same issues again - if you have a record of how you previously solved a problem, looking back through this journal should help you. You are also expected to answer any <tt class="docutils literal"><span class="pre">Extra</span> <span class="pre">Credit</span></tt> questions that are raised in the LPTHW text in your programming notebook. I don&#8217;t, however, need to see all of the code snippets that you write for the exercises here.</p>
<p>I recommend that you keep notes in whatever program you&#8217;re comfortable with, provided that when I collect these you&#8217;ll be able to generate a PDF that you can send to me. Some choices might be MS Word, Google Docs, Evernote, Gedit (you&#8217;ll need to print to PDF), etc.</p>
<p>A few (random) times through-out the semester I&#8217;ll collect your programming notebooks in electronic form. I&#8217;ll review these and score them on completeness.</p>
<p>Here&#8217;s an example of what your notebook might look like for the first three exercises:</p>
<div class="highlight-python"><pre>Exercise 0 (31 Aug 2011)

Installed Gedit and created a test file. Had to look up how to make and
change directories on the web. I googled "how to make a directory in the
Terminal" and found this page: http://guides.macrumors.com/Terminal. This
pages covers some useful commands for working the in terminal - I found
mkdir and cd here.

Exercise 1

This exercise was easy. Ran into one issue where I got a syntax error
because I used a double-quote to start a string and a single-quote to
end a string. Easy fix once I figured out the problem.

Extra credits: 1) Just added another print statement; 2) Deleted all of
the lines except for one; 3) The '#' symbol causes a line not to be
interpreted - if you add this to the line everything after the # is ignored.
I figured this out by experimenting with putting # signs in my code and
rerunning it. This would have been a better way to achieve the result of
exercise (2).

Exercise 2 (1 Sept 2011)

No problems with writing or running the code.

Extra credits: 1) Yep, I was right about what the '#' symbol does. 3) Missed a
space between print and "This will run." on line 9 when re-entering the code. This
caused a syntax error.</pre>
</div>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        «&#160;&#160;<a href="books_and_links.html">Books and recommended references</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="application_presentations.html">Application presentations</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer">
        &copy; Copyright 2012, J Gregory Caporaso.
      Last updated on Feb 12, 2014.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>
