

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Lecture 10: Sequence alignment algorithms (continued) &mdash; bioinformatics@nau</title>
    
    <link rel="stylesheet" href="../../../_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/print.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../../_static/theme_extras.js"></script>
    <link rel="top" title="bioinformatics@nau" href="../../../index.html" />
    <link rel="up" title="Lecture outlines" href="index.html" />
    <link rel="next" title="Lecture 11: Alignments and phylogenetic trees" href="lecture11.html" />
    <link rel="prev" title="Lecture 9: Sequence alignment algorithms" href="lecture9.html" /> 
  </head>
  <body>
      <div class="header"><h1 class="heading"><a href="../../../index.html">
          <span>bioinformatics@nau</span></a></h1>
        <h2 class="heading"><span>Lecture 10: Sequence alignment algorithms (continued)</span></h2>
      </div>
      <div class="topnav">
      
        <p>
        «&#160;&#160;<a href="lecture9.html">Lecture 9: Sequence alignment algorithms</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="lecture11.html">Lecture 11: Alignments and phylogenetic trees</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="lecture-10-sequence-alignment-algorithms-continued">
<h1>Lecture 10: Sequence alignment algorithms (continued)<a class="headerlink" href="#lecture-10-sequence-alignment-algorithms-continued" title="Permalink to this headline">¶</a></h1>
<p>The following coding examples will cover the various features and tools in python that you&#8217;ve learned about (or will very shortly) and how they can be applied to implement the Needleman-Wunsch alignment algorithm. This same lesson can be applied to the Smith-Waterman alignment algorithm.</p>
<div class="section" id="two-dimensional-lists">
<h2>Two dimensional lists<a class="headerlink" href="#two-dimensional-lists" title="Permalink to this headline">¶</a></h2>
<p>You&#8217;ve been introduced to lists in python, but it&#8217;s common that you need to represent multi-dimensional data such as a matrix. To do this, you can use a list where each entry in the list is a list. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">matrix</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">],[</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">]]</span>
</pre></div>
</div>
<p>You can then access entries in this list by indicating their index in order from the top-most list to the deeper lists. For example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># access the second entry of the second list</span>
<span class="n">matrix</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>Try accessing some different values. You can also access one of the full inner lists in the usual way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">matrix</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="section" id="dictionaries">
<h2>Dictionaries<a class="headerlink" href="#dictionaries" title="Permalink to this headline">¶</a></h2>
<p>A key concept is introduced in LPTHW exercise 40: the dictionary or <tt class="docutils literal"><span class="pre">dict</span></tt>. These allow for efficient look-ups of <cite>values</cite> associated with a <cite>key</cite>, and are used constantly in python.</p>
<p>Some of the key features associated with dictionaries include creating dictionaries, setting and resetting the values for certain keys, and looking up the values of certain keys. Dictionaries are created as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">base_comps</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span> <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span> <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span> <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">}</span>
</pre></div>
</div>
<p>This <tt class="docutils literal"><span class="pre">dict</span></tt> would be said to <cite>map</cite> each of the four DNA bases to its complement. You can do <cite>lookups</cite> on this dictionary as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">base_comps</span><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>If you want to add or reset a value after dictionary creation you can do so as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># add a lookup for uracil</span>
<span class="n">base_comps</span><span class="p">[</span><span class="s">&#39;U&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;A&#39;</span>
<span class="c"># remap the lookup for adenine</span>
<span class="n">base_comps</span><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s">&#39;U&#39;</span>
<span class="c"># delete the entry for thymine</span>
<span class="k">del</span> <span class="n">base_comps</span><span class="p">[</span><span class="s">&#39;T&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>You can also check whether a key is contained in a dictionary using <tt class="docutils literal"><span class="pre">in</span></tt>. For example, try these:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="s">&#39;A&#39;</span> <span class="ow">in</span> <span class="n">base_comps</span>

<span class="s">&#39;W&#39;</span> <span class="ow">in</span> <span class="n">base_comps</span>
</pre></div>
</div>
<p>You&#8217;ll notice that if you access an entry that is not in a dictionary you&#8217;ll get a <tt class="docutils literal"><span class="pre">KeyError</span></tt>. The error messages that you get when a <tt class="docutils literal"><span class="pre">KeyError</span></tt> is <cite>thrown</cite> may not always be very intuitive so adding a little additional error handling is a good idea. One way to do this is with an if/else statement, where you check for the presence of a key in a dictionary:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">base_comps</span><span class="p">:</span>
    <span class="k">return</span> <span class="n">base_comps</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
<span class="k">else</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s"> is not in the base complement lookup dictionary&quot;</span> <span class="o">%</span> <span class="n">key</span>
    <span class="nb">exit</span><span class="p">()</span>
</pre></div>
</div>
<p>More commonly, expert python users will <cite>catch</cite> and handle the error if it happens using a <tt class="docutils literal"><span class="pre">try/except</span></tt>. The following is a very common python idiom, so I want to introduce it in case you run into it in other code. Don&#8217;t worry about integrating into your code too much for now - wait until you learn a little about <tt class="docutils literal"><span class="pre">Exceptions</span></tt> in python:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">try</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">base_comps</span><span class="p">[</span><span class="n">key</span><span class="p">]</span>
<span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
    <span class="k">print</span> <span class="s">&quot;</span><span class="si">%s</span><span class="s"> is not in the base complement lookup dictionary&quot;</span> <span class="o">%</span> <span class="n">key</span>
    <span class="nb">exit</span><span class="p">()</span>
</pre></div>
</div>
<div class="section" id="using-a-dictionary-to-simplify-our-reverse-complementing-code">
<h3>Using a dictionary to simplify our reverse complementing code<a class="headerlink" href="#using-a-dictionary-to-simplify-our-reverse-complementing-code" title="Permalink to this headline">¶</a></h3>
<p>Here&#8217;s an example that will simplify our reverse complementing code:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="kn">from</span> <span class="nn">sys</span> <span class="kn">import</span> <span class="n">argv</span>
<span class="n">script_name</span><span class="p">,</span> <span class="n">sequence</span> <span class="o">=</span> <span class="n">argv</span>

<span class="n">complements</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="s">&#39;T&#39;</span><span class="p">,</span>
               <span class="s">&#39;G&#39;</span><span class="p">:</span><span class="s">&#39;C&#39;</span><span class="p">,</span>
               <span class="s">&#39;T&#39;</span><span class="p">:</span><span class="s">&#39;A&#39;</span><span class="p">,</span>
               <span class="s">&#39;C&#39;</span><span class="p">:</span><span class="s">&#39;G&#39;</span><span class="p">}</span>

<span class="k">def</span> <span class="nf">reverse_complement_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">,</span><span class="n">complements</span><span class="o">=</span><span class="n">complements</span><span class="p">):</span>
    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c"># note that we&#39;re iterating over sequence in reverse here</span>
    <span class="k">for</span> <span class="n">base</span> <span class="ow">in</span> <span class="n">sequence</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
        <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">complements</span><span class="p">[</span><span class="n">base</span><span class="p">])</span>
    <span class="k">return</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="n">rc_sequence</span> <span class="o">=</span> <span class="n">reverse_complement_sequence</span><span class="p">(</span><span class="n">sequence</span><span class="p">)</span>
<span class="k">print</span> <span class="n">rc_sequence</span>
</pre></div>
</div>
</div>
<div class="section" id="implementing-substitution-matrices-with-dictionaries">
<h3>Implementing substitution matrices with dictionaries<a class="headerlink" href="#implementing-substitution-matrices-with-dictionaries" title="Permalink to this headline">¶</a></h3>
<p>As with lists, you can create nested dictionaries, where each entry in the dictionary is a dictionary. One application of this is defining substitution matrices for use in sequence alignment. If you wanted to score different base substitutions differently, for example, you could define a substitution matrix as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">substitution_matrix</span> <span class="o">=</span> <span class="p">{</span><span class="s">&#39;A&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;T&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">},</span>
                       <span class="s">&#39;C&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;T&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">},</span>
                       <span class="s">&#39;G&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span><span class="s">&#39;T&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">},</span>
                       <span class="s">&#39;T&#39;</span><span class="p">:{</span><span class="s">&#39;A&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;C&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">2</span><span class="p">,</span><span class="s">&#39;G&#39;</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="s">&#39;T&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}}</span>
</pre></div>
</div>
<p>This scoring matrix is scoring all substitutions events with negative scores, but differentiating between <cite>transition</cite> and <cite>transversion</cite> substitutions. If these terms are unfamiliar, look them up. Why might you score these differently?</p>
<p>A commonly used scoring matrix for protein alignments is the BLOSUM50. To define the BLOSUM50 substitution matrix as a dictionary you could run the following:</p>
<div class="highlight-python"><pre>   blosum50 = {'A': {'A': 5.0, 'C': -1.0, 'D': -2.0, 'E': -1.0, 'F': -3.0, 'G': 0.0, 'H': -2.0, 'I': -1.0, 'K': -1.0, 'L': -2.0, 'M': -1.0, 'N': -1.0, 'P': -1.0, 'Q': -1.0, 'R': -2.0, 'S': 1.0, 'T': 0.0, 'V': 0.0, 'W': -3.0, 'Y': -2.0},
'C': {'A': -1.0, 'C': 13.0, 'D': -4.0, 'E': -3.0, 'F': -2.0, 'G': -3.0, 'H': -3.0, 'I': -2.0, 'K': -3.0, 'L': -2.0, 'M': -2.0, 'N': -2.0, 'P': -4.0, 'Q': -3.0, 'R': -4.0, 'S': -1.0, 'T': -1.0, 'V': -1.0, 'W': -5.0, 'Y': -3.0},
'D': {'A': -2.0, 'C': -4.0, 'D': 8.0, 'E': 2.0, 'F': -5.0, 'G': -1.0, 'H': -1.0, 'I': -4.0, 'K': -1.0, 'L': -4.0, 'M': -4.0, 'N': 2.0, 'P': -1.0, 'Q': 0.0, 'R': -2.0, 'S': 0.0, 'T': -1.0, 'V': -4.0, 'W': -5.0, 'Y': -3.0},
'E': {'A': -1.0, 'C': -3.0, 'D': 2.0, 'E': 6.0, 'F': -3.0, 'G': -3.0, 'H': 0.0, 'I': -4.0, 'K': 1.0, 'L': -3.0, 'M': -2.0, 'N': 0.0, 'P': -1.0, 'Q': 2.0, 'R': 0.0, 'S': -1.0, 'T': -1.0, 'V': -3.0, 'W': -3.0, 'Y': -2.0},
'F': {'A': -3.0, 'C': -2.0, 'D': -5.0, 'E': -3.0, 'F': 8.0, 'G': -4.0, 'H': -1.0, 'I': 0.0, 'K': -4.0, 'L': 1.0, 'M': 0.0, 'N': -4.0, 'P': -4.0, 'Q': -4.0, 'R': -3.0, 'S': -3.0, 'T': -2.0, 'V': -1.0, 'W': 1.0, 'Y': 4.0},
'G': {'A': 0.0, 'C': -3.0, 'D': -1.0, 'E': -3.0, 'F': -4.0, 'G': 8.0, 'H': -2.0, 'I': -4.0, 'K': -2.0, 'L': -4.0, 'M': -3.0, 'N': 0.0, 'P': -2.0, 'Q': -2.0, 'R': -3.0, 'S': 0.0, 'T': -2.0, 'V': -4.0, 'W': -3.0, 'Y': -3.0},
'H': {'A': -2.0, 'C': -3.0, 'D': -1.0, 'E': 0.0, 'F': -1.0, 'G': -2.0, 'H': 10.0, 'I': -4.0, 'K': 0.0, 'L': -3.0, 'M': -1.0, 'N': 1.0, 'P': -2.0, 'Q': 1.0, 'R': 0.0, 'S': -1.0, 'T': -2.0, 'V': -4.0, 'W': -3.0, 'Y': 2.0},
'I': {'A': -1.0, 'C': -2.0, 'D': -4.0, 'E': -4.0, 'F': 0.0, 'G': -4.0, 'H': -4.0, 'I': 5.0, 'K': -3.0, 'L': 2.0, 'M': 2.0, 'N': -3.0, 'P': -3.0, 'Q': -3.0, 'R': -4.0, 'S': -3.0, 'T': -1.0, 'V': 4.0, 'W': -3.0, 'Y': -1.0},
'K': {'A': -1.0, 'C': -3.0, 'D': -1.0, 'E': 1.0, 'F': -4.0, 'G': -2.0, 'H': 0.0, 'I': -3.0, 'K': 6.0, 'L': -3.0, 'M': -2.0, 'N': 0.0, 'P': -1.0, 'Q': 2.0, 'R': 3.0, 'S': 0.0, 'T': -1.0, 'V': -3.0, 'W': -3.0, 'Y': -2.0},
'L': {'A': -2.0, 'C': -2.0, 'D': -4.0, 'E': -3.0, 'F': 1.0, 'G': -4.0, 'H': -3.0, 'I': 2.0, 'K': -3.0, 'L': 5.0, 'M': 3.0, 'N': -4.0, 'P': -4.0, 'Q': -2.0, 'R': -3.0, 'S': -3.0, 'T': -1.0, 'V': 1.0, 'W': -2.0, 'Y': -1.0},
'M': {'A': -1.0, 'C': -2.0, 'D': -4.0, 'E': -2.0, 'F': 0.0, 'G': -3.0, 'H': -1.0, 'I': 2.0, 'K': -2.0, 'L': 3.0, 'M': 7.0, 'N': -2.0, 'P': -3.0, 'Q': 0.0, 'R': -2.0, 'S': -2.0, 'T': -1.0, 'V': 1.0, 'W': -1.0, 'Y': 0.0},
'N': {'A': -1.0, 'C': -2.0, 'D': 2.0, 'E': 0.0, 'F': -4.0, 'G': 0.0, 'H': 1.0, 'I': -3.0, 'K': 0.0, 'L': -4.0, 'M': -2.0, 'N': 7.0, 'P': -2.0, 'Q': 0.0, 'R': -1.0, 'S': 1.0, 'T': 0.0, 'V': -3.0, 'W': -4.0, 'Y': -2.0},
'P': {'A': -1.0, 'C': -4.0, 'D': -1.0, 'E': -1.0, 'F': -4.0, 'G': -2.0, 'H': -2.0, 'I': -3.0, 'K': -1.0, 'L': -4.0, 'M': -3.0, 'N': -2.0, 'P': 10.0, 'Q': -1.0, 'R': -3.0, 'S': -1.0, 'T': -1.0, 'V': -3.0, 'W': -4.0, 'Y': -3.0},
'Q': {'A': -1.0, 'C': -3.0, 'D': 0.0, 'E': 2.0, 'F': -4.0, 'G': -2.0, 'H': 1.0, 'I': -3.0, 'K': 2.0, 'L': -2.0, 'M': 0.0, 'N': 0.0, 'P': -1.0, 'Q': 7.0, 'R': 1.0, 'S': 0.0, 'T': -1.0, 'V': -3.0, 'W': -1.0, 'Y': -1.0},
'R': {'A': -2.0, 'C': -4.0, 'D': -2.0, 'E': 0.0, 'F': -3.0, 'G': -3.0, 'H': 0.0, 'I': -4.0, 'K': 3.0, 'L': -3.0, 'M': -2.0, 'N': -1.0, 'P': -3.0, 'Q': 1.0, 'R': 7.0, 'S': -1.0, 'T': -1.0, 'V': -3.0, 'W': -3.0, 'Y': -1.0},
'S': {'A': 1.0, 'C': -1.0, 'D': 0.0, 'E': -1.0, 'F': -3.0, 'G': 0.0, 'H': -1.0, 'I': -3.0, 'K': 0.0, 'L': -3.0, 'M': -2.0, 'N': 1.0, 'P': -1.0, 'Q': 0.0, 'R': -1.0, 'S': 5.0, 'T': 2.0, 'V': -2.0, 'W': -4.0, 'Y': -2.0},
'T': {'A': 0.0, 'C': -1.0, 'D': -1.0, 'E': -1.0, 'F': -2.0, 'G': -2.0, 'H': -2.0, 'I': -1.0, 'K': -1.0, 'L': -1.0, 'M': -1.0, 'N': 0.0, 'P': -1.0, 'Q': -1.0, 'R': -1.0, 'S': 2.0, 'T': 5.0, 'V': 0.0, 'W': -3.0, 'Y': -2.0},
'V': {'A': 0.0, 'C': -1.0, 'D': -4.0, 'E': -3.0, 'F': -1.0, 'G': -4.0, 'H': -4.0, 'I': 4.0, 'K': -3.0, 'L': 1.0, 'M': 1.0, 'N': -3.0, 'P': -3.0, 'Q': -3.0, 'R': -3.0, 'S': -2.0, 'T': 0.0, 'V': 5.0, 'W': -3.0, 'Y': -1.0},
'W': {'A': -3.0, 'C': -5.0, 'D': -5.0, 'E': -3.0, 'F': 1.0, 'G': -3.0, 'H': -3.0, 'I': -3.0, 'K': -3.0, 'L': -2.0, 'M': -1.0, 'N': -4.0, 'P': -4.0, 'Q': -1.0, 'R': -3.0, 'S': -4.0, 'T': -3.0, 'V': -3.0, 'W': 15.0, 'Y': 2.0},
'Y': {'A': -2.0, 'C': -3.0, 'D': -3.0, 'E': -2.0, 'F': 4.0, 'G': -3.0, 'H': 2.0, 'I': -1.0, 'K': -2.0, 'L': -1.0, 'M': 0.0, 'N': -2.0, 'P': -3.0, 'Q': -1.0, 'R': -1.0, 'S': -2.0, 'T': -2.0, 'V': -1.0, 'W': 2.0, 'Y': 8.0}}</pre>
</div>
<p>After defining this, you can look up scores in the 2D dictionary as follows:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">print</span> <span class="n">blosum50</span><span class="p">[</span><span class="s">&#39;Y&#39;</span><span class="p">][</span><span class="s">&#39;W&#39;</span><span class="p">]</span>
<span class="k">print</span> <span class="n">blosum50</span><span class="p">[</span><span class="s">&#39;D&#39;</span><span class="p">][</span><span class="s">&#39;W&#39;</span><span class="p">]</span>
<span class="k">print</span> <span class="n">blosum50</span><span class="p">[</span><span class="s">&#39;A&#39;</span><span class="p">][</span><span class="s">&#39;A&#39;</span><span class="p">]</span>
</pre></div>
</div>
<p>We&#8217;ll use this matrix to score some alignments and to explore the Needleman-Wunsch and Smith-Waterman alignment algorithms.</p>
</div>
</div>
<div class="section" id="nested-for-loops">
<h2>Nested for loops<a class="headerlink" href="#nested-for-loops" title="Permalink to this headline">¶</a></h2>
<p>Sometimes one for loop isn&#8217;t enough. For example, maybe you have you two lists and you need to perform some operation on all pairs of elements in those two lists. One application of this is in sequence alignment, where we need to compare all pairs of positions to compute alignment scores. In this example we&#8217;ll look at how that might be computed.</p>
<p>Imagine we&#8217;re starting with two sequences and we want to compute a scoring matrix to use for aligning those two sequences:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># Define the sequences we want to align</span>
<span class="n">seq1</span> <span class="o">=</span> <span class="s">&quot;HEAGAWGHEE&quot;</span>
<span class="n">seq2</span> <span class="o">=</span> <span class="s">&quot;PAWHEAE&quot;</span>

<span class="c"># Initialize a matrix to use for storing the scores</span>
<span class="n">match_matrix</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c"># Iterate over the amino acids in sequence two (which will correspond</span>
<span class="c"># to the vertical sequence in the matrix)</span>
<span class="k">for</span> <span class="n">aa2</span> <span class="ow">in</span> <span class="n">seq2</span><span class="p">:</span>
    <span class="c"># Initialize the current row of the matrix</span>
    <span class="n">current_row</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="c"># Iterate over the amino acids in sequence one (which will</span>
    <span class="c"># correspond to the horizontal sequence in the matrix)</span>
    <span class="k">for</span> <span class="n">aa1</span> <span class="ow">in</span> <span class="n">seq1</span><span class="p">:</span>
        <span class="c"># score each position in the matrix according to it&#39;s value in</span>
        <span class="c"># blosum50</span>
        <span class="n">current_row</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">blosum50</span><span class="p">[</span><span class="n">aa1</span><span class="p">][</span><span class="n">aa2</span><span class="p">])</span>
    <span class="c"># append the current row to the matrix</span>
    <span class="n">match_matrix</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">current_row</span><span class="p">)</span>
</pre></div>
</div>
<p>You can now print out <tt class="docutils literal"><span class="pre">match_matrix</span></tt>. You can do that in a simply way with the following code:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">match_matrix</span><span class="p">:</span>
    <span class="k">print</span> <span class="n">row</span>
</pre></div>
</div>
<p>Or you can get fancy and print it in a prettier way:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="c"># define a format string that will be used to format each line -</span>
<span class="c"># since seq1 is the &#39;horizontal&#39; sequence in our matrix we&#39;ll</span>
<span class="c"># have len(seq1) + 1 entries on each line to print the scores and</span>
<span class="c"># seq2 base associated with each row</span>
<span class="n">line_format</span> <span class="o">=</span> <span class="s">&quot;</span><span class="si">%6s</span><span class="s">&quot;</span> <span class="o">*</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

<span class="c"># print seq1 (start the line with an empty string)</span>
<span class="k">print</span> <span class="n">line_format</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">([</span><span class="s">&#39; &#39;</span><span class="p">]</span> <span class="o">+</span> <span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="nb">list</span><span class="p">(</span><span class="n">seq1</span><span class="p">)))</span>

<span class="c"># iterate over the rows and print each (starting with the</span>
<span class="c"># corresponding base in sequence2)</span>
<span class="k">for</span> <span class="n">row</span><span class="p">,</span> <span class="n">base</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">match_matrix</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
    <span class="k">print</span> <span class="n">line_format</span> <span class="o">%</span> <span class="nb">tuple</span><span class="p">([</span><span class="n">base</span><span class="p">]</span> <span class="o">+</span> <span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span><span class="n">row</span><span class="p">))</span>
</pre></div>
</div>
<p>The formatting sting <tt class="docutils literal"><span class="pre">%6s</span></tt> prints a string using exactly six characters (padded with spaces), so that allows us to nicely align the columns.</p>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        «&#160;&#160;<a href="lecture9.html">Lecture 9: Sequence alignment algorithms</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="lecture11.html">Lecture 11: Alignments and phylogenetic trees</a>&#160;&#160;»
        </p>

      </div>

    <div class="footer">
        &copy; Copyright 2012, J Gregory Caporaso.
      Last updated on Jan 28, 2013.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>